### ISSUE TITLE
Bug: several versions of the actionview RubyGem >= 4.2.0 are incorrectly listed as vulnerable to CVE-2016-2097

### ISSUE THREAD

**Issue by csansone-handy** [OSSIndex/vulns#44]
**Example Vulnerability URLs**
```
https://ossindex.sonatype.org/vuln/63e3ccea-cae4-4ba2-b8f7-06633d4587b0
```

**Description**
Description
Several versions of the actionview RubyGem that are version >= 4.2.0 are incorrectly listed as being vulnerable to CVE-2016-2098.

According to [this Rubysec advisory](CVE-2016-2097) the following versions of actionview are **NOT** vulnerable: ~> 4.1.14, >= 4.1.14.2, >= 4.2.0

Can you please update this in OSS Index?

Furthermore, is it also possible to elaborate what the version range on the vulnerability URL is supposed to represent? Is this supposed to represent the vulnerable versions or the safe/patched versions? As it stands now the page seems to be listing the clean/non-vulnerable versions as opposed to the vulnerable versions:

![image](https://user-images.githubusercontent.com/51127567/68126258-65c4be00-fee1-11e9-9acb-4a912d6e33b0.png)

I am wondering if there is a larger issue where some logic regarding the version range is inverted causing the non-vulnerable versions to be listed as vulnerable since we had a similiar issue with actionpack as I reported [here](https://github.com/OSSIndex/vulns/issues/27).

**Comment by csansone-handy** [comment]
Anyone have a chance to look into this?

**Comment by ken-duck** [comment]
Sorry for the delay. I am resolving the issue now and will have the fix in hopefully by tomorrow (it takes a bit for data to move down the pipe). I have been falling behind on a few things lately. We have some additional team members joining shortly which will help reduce the load and hopefully we will see less lag.

**Comment by csansone-handy** [comment]
Thanks Ken. Do you think this is a one-off issue? Will anyone be looking into the matching logic for Ruby advisories to make sure there isn't a larger issue?

**Comment by ken-duck** [comment]
It will take me a while to root this one out; I need to figure out the source of the data.

I will definitely figure out whether it is a code problem. I have a suspicion it might be human error in this case, but the fact that you have found two of them worries me.

**Comment by ken-duck** [comment]
FYI, it is definitely a larger issue. We pull the version ranges automatically from rubysec for some vulnerabilities, and it appears that conversion from rubysec ranges does not work in all cases. I am exploring the issue now to try and figure out where the problem is. I have fixed this one by hand, but have a sneaking suspicion it may be getting overwritten. If it is, then I plan on fixing *that* problem before the end of Tuesday, and trying to resolve the parsing issue shortly thereafter.

The exciting part is trying to make sense of this:

> unaffected_versions:
>   - ">= 4.2.0"
> 
> patched_versions:
>   - "~> 4.1.14, >= 4.1.14.2"

To me that does not read properly. I need to try and understand what that means, then attempt to reasonably parse it.

I have not figured out how the code decided on the version range you actually saw. I think perhaps that was an earlier version of the code which had a different bug and was superseded, because the current code actually fails to parse the patched version.

Anyhow, you definitely identified a problem and I am endeavouring to figure out an appropriate approach forward now.

**Comment by csansone-handy** [comment]
Thanks for digging in Ken. I would read it as:

> If you are running any version that is listed in **EITHER** the `unaffected_versions` **OR** the `patched_versions` then this vulnerability does **NOT** affect you. However, if you are running any other version of the library then you are indeed vulnerable.

In this specific example it would mean if you are running any version greater than 4.2.0 (`>= 4.2.0`) OR if you are running version 4.1.14 (`~> 4.1.14`) OR if you are running any version greater than 4.1.14.2 (`>= 4.1.14.2`) then you are NOT affected by the vulnerability. Any other version is affected.

It might be best to try and sync up with someone from RubySec to confirm the above.

**Comment by ken-duck** [comment]
I think you are more or less right, except that the ranges `~>4.1.14` and `>= 4.1.14.2` intersect (as `~4.1.14` is equivalent to `>=4.1.14 <4.1.15`.

The text in the description is much easier to understand for me:

```
  Versions Affected:  3.2.x, 4.0.x, 4.1.x
  Not affected:       4.2+
  Fixed Versions:     3.2.22.2, 4.1.14.2
```

It looks like it might be a typo. I will bring it up with the RubySec folks as you suggest.

**Comment by ken-duck** [comment]
I did a visual scan of version ranges for other similar vulnerabilities in OSS Index, as well as a few ad-hoc queries against the database and did not find any others that appear to be grotesquely incorrect (in particular, there are no more open ended ranges that do not appear to be on purpose).

I raised an issue with the RubySec folks regarding the range for the issue above (https://github.com/rubysec/ruby-advisory-db/issues/427).

Meanwhile, the specific issue you wrote about appears to be fixed and holding true (not getting ovrewritten). I will monitor for a few more days.

I am pondering the range parsing algorithm some more to see if there are easy improvements to make. It does not look like it should cause the strange issue you detected anymore. It is possible there are other weaknesses in the algorithm, though, so I will explore it a bit more to see if anything springs to mind.
